File: 동적 계획법 1/12015(가장 긴 증가하는 부분 수열 2)/12015(가장 긴 증가하는 부분 수열 2).cpp
#include <cstdio>
#include <vector>
#define MAXSIZE 1000000
using namespace std;

int input[MAXSIZE] = { 0 };

int main(void)
{
	vector<int> answer;
	int N;

	answer.reserve(MAXSIZE);
	scanf("%d", &N);
	for (int i = 0; i < N; i++) scanf("%d", &input[i]);
	for (int i = 0; i < N; i++) 
	{
		int s, e;
		int val = input[i];

		s = 0;
		e = answer.size() - 1;
		while (s <= e)
		{
			int mid = (s + e) / 2;
			
			if (answer[mid] < val)
				s = mid + 1;
			else if (answer[mid] > val)
				e = mid - 1;
			else
				break;
		}
		if (s >= answer.size())
			answer.push_back(val);
		else
			answer[s] = val;
	}
	printf("%d\n", answer.size());
	return 0;
}

File: 동적 계획법 1/12015(가장 긴 증가하는 부분 수열 2)/12015(가장 긴 증가하는 부분 수열 2).py
import sys
input = sys.stdin.readline

N = int(input())
arr = list(map(int, input().split()))
answer = []
for val in arr:
    if len(answer) == 0:
        answer.append(val)
    else:
        if answer[-1] < val:
            answer.append(val)
        elif answer[-1] > val:
            s = 0
            e = len(answer) - 1
            while s <= e:
                mid = (s + e) // 2
                if answer[mid] == val:
                    e = mid
                    break
                elif answer[mid] > val:
                    e = mid - 1
                elif answer[mid] < val:
                    s = mid + 1
            if answer[e] == val:
                continue
            else:
                answer[e + 1] = val

print(len(answer))
